src/node_api.cc

```diff
@@ -47,26 +48,38 @@ void node_napi_env__::CallFinalizer(napi_finalize cb, void* data, void* hint) {
 
 template <bool enforceUncaughtExceptionPolicy>
 void node_napi_env__::CallFinalizer(napi_finalize cb, void* data, void* hint) {
-  if (destructing) {
-    // we can not defer finalizers when the environment is being destructed.
-    v8::HandleScope handle_scope(isolate);
-    v8::Context::Scope context_scope(context());
-    CallbackIntoModule<enforceUncaughtExceptionPolicy>(
-        [&](napi_env env) { cb(env, data, hint); });
-    return;
+  v8::HandleScope handle_scope(isolate);
+  v8::Context::Scope context_scope(context());
+  CallbackIntoModule<enforceUncaughtExceptionPolicy>(
+      [&](napi_env env) { cb(env, data, hint); });
+}
+
+void node_napi_env__::EnqueueFinalizer(v8impl::RefTracker* finalizer) {
+  napi_env__::EnqueueFinalizer(finalizer);
+  // Schedule a second pass only when it has not been scheduled, and not
+  // destructing the env.
+  // When the env is being destructed, queued finalizers are drained in the
+  // loop of `node_napi_env__::DrainFinalizerQueue`.
+  if (!finalization_scheduled && !destructing) {
+    finalization_scheduled = true;
+    Ref();
+    node_env()->SetImmediate([this](node::Environment* node_env) {
+      finalization_scheduled = false;
+      Unref();
```

Comment:
If it's possible that Finalize on line 81 below could invoke JavaScript should this Unref be after DrainFinalizerQueue()  to make sure the env stays alive until it is complete?

Reply:
I may have answered my own question, I can see that BufferFinalizer and ThreadSafeFunctions keep their own Ref() on the env, so an additional ref while the queue is processed is unecessary.